components/region_cache_memory_engine/src/background.rs

```diff
@@ -451,6 +603,76 @@ impl BackgroundRunnerCore {
         #[cfg(test)]
         flush_epoch();
     }
+
+    fn evict_on_soft_limit_reached(&mut self) {
+        if self.range_stats_manager.is_none() {
+            warn!("range stats manager is not initialized, cannot evict on soft limit reached");
+            return;
+        }
+        let range_stats_manager = self.range_stats_manager.as_ref().unwrap();
+        let to_shrink_by = self
+            .memory_controller
+            .mem_usage()
+            .checked_sub(self.memory_controller.soft_limit_threshold());
+        if to_shrink_by.is_none() {
+            return;
+        }
+        let mut remaining = to_shrink_by.unwrap();
+        let mut ranges_to_evict = Vec::<(CacheRange, u64)>::with_capacity(256);
+        range_stats_manager.collect_candidates_for_eviction(&mut ranges_to_evict);
+        let ranges_to_evict = ranges_to_evict
+            .iter()
+            .filter(|(range, _)| self.engine.read().range_manager().contains_range(range))
+            .rev();
+        for (range, approx_size) in ranges_to_evict {
+            if remaining == 0 {
+                break;
+            }
+            if self.engine.write().mut_range_manager().evict_range(range) {
+                info!("evict on soft limit reached"; "range" => ?&range, "approx_size" => approx_size, "remaining" => remaining);
+                remaining = remaining
+                    .checked_sub(*approx_size as usize)
+                    .unwrap_or_default();
+                range_stats_manager.handle_range_evicted(range);
+            }
+        }
+    }
+
+    fn top_regions_load_evict(&self) {
```

Comment:
TODO: document and an integration/unit-level test. Hard to do this precisely, failed a few times, will do again.